Implement conditional workflow steps

mvdbeek opened the pull request:
This PR implements executing tool and subworkflow steps conditionally, based on the result of a javascript expression. The workflow editor will for now only allow the simplest expression, which is a connection to a boolean parameter (from an expression tool or a workflow input).
An example gxformat2 workflow looks like this:
```yaml
class: GalaxyWorkflow
inputs:
  should_run:
    type: boolean
  some_file:
    type: data
steps:
  cat1:
    tool_id: cat1
    in:
      input1: some_file
      should_run: should_run
    when: $(inputs.should_run)
```
If `when` evaluates to `true` the step will run, if it evaluates to `false` it will not run.
This is also implemented for mapped over tool and workflow steps; if an output is skipped it will be produce a null value. You can think of this as an optional output. That means you can use the `pickValue` expression tool to pick for instance the first created output for two mutually exclusive processing steps.

Includes a whole new infrastructure for displaying scheduling warnings, failures and cancellations:

![Screenshot 2023-01-16 at 22 23 19](https://user-images.githubusercontent.com/6804901/212771075-7d4f5f00-a428-461d-b618-cb9f2cfb0d75.png)


Remaining work: 
 - Update the workflow editor
 - Some visual indicator for partial collections in the history ?

## How to test the changes?
(Select all options that apply)
- [x] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [ ] This is a refactoring of components with existing test coverage.
- [ ] Instructions for manual testing are as follows:
  1. [add testing steps and prerequisites here if you didn't write automated tests covering all your changes]

## License
- [x] I agree to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT).

jmchilton committed:
Implement conditionals via boolean step parameter

For tool steps only right now. This side-steps user-defined javascript
expressions in favor of boolean input parameters.

mvdbeek committed:
Attempt to skip mapped-over steps

mvdbeek committed:
Into map-over

mvdbeek committed:
Add when_expression migration

mvdbeek committed:
WIP: test map over and pick_value

mvdbeek committed:
Fix up migration

mvdbeek committed:
WIP: Produce json null on skip, load expression.json null as null in json_wrap

Should probably avoid the object store assignment here, but I would also
like to avoid the extra pass through the job handler. OTOH a skipped job
might be nice ?

mvdbeek committed:
Fix up map_over of conditional output

mvdbeek committed:
Assert skipped state in tests

mvdbeek committed:
Drop unreachable SkipWorkflowStepEvaluation except statement

mvdbeek committed:
Drop unused function

mvdbeek committed:
Fix type annotation for temp_input_connections

mvdbeek committed:
Make new tests robust to job and step order

which seems to not be stable?!

mvdbeek committed:
Make sure we're not changing datatypes on skipped outputs

mvdbeek committed:
Implement conditional subwokflow steps

for mapped over subworkflows and simple subworkflows

mvdbeek committed:
Add skipped to job state summary

mvdbeek committed:
Stick with CWL syntax and CWL tooling to implement step skipping

mvdbeek committed:
Add subworkflow skip test when tool structure changes

mvdbeek committed:
Fix mapping over subworkflow steps without explicit connection to outer workflow

mvdbeek committed:
Mark skipped outputs as populated

mvdbeek committed:
Test and fix conditional subworkflow steps where subworkflow step produces additional mapping over without being directly connected to a workflow input

jmchilton commented on lib/galaxy/workflow/modules.py:
```diff
@@ -103,6 +105,99 @@ def __str__(self):
+
+    hda_references = []
+
+    def to_cwl(value):
```
This is so complicated and we're only basically testing boolean parameters right?

What if an integer is sent - what happens? A dataset? Etc...

Can we have a test that the workflow dies if an invalid expression is set in there?

mvdbeek replied:
I've included a bunch of tests and I've also added "invocation messages", see the screenshot in the description ...  I think we're now displaying (almost?) all of the possible workflow errors in the API and UI.

mvdbeek committed:
Fix migration down_revision

mvdbeek committed:
Update typescript schema

mvdbeek committed:
Backend for showing invocation failure/cancellation/warning reasons

mvdbeek committed:
Record execution_tracker.execution_errors on invocation

mvdbeek committed:
Test invalid expressions and expression results

mvdbeek committed:
Avoid Union of GenericModel

More verbose, but https://github.com/pydantic/pydantic/issues/1194
prevents a more concise setup.

mvdbeek committed:
Add migration for updated HistoryDatasetCollectionJobStateSummary

mvdbeek committed:
Move InvocationMessage models to own file

mvdbeek committed:
Generate model for InvocationMessage

mvdbeek committed:
Invocation schema import fixes

mvdbeek committed:
Fix dependent workflow step id

mvdbeek committed:
Create and use useWorkflowInstance composable

mvdbeek committed:
Make workflow_step_id required for workflow_output_not_found warning

mvdbeek committed:
Fix regular workflow file upload in UI

mvdbeek committed:
Add InvocationMessage component and mount in invocation summary tab

jmchilton commented on lib/galaxy/workflow/modules.py:
```diff
@@ -104,6 +118,126 @@ def __str__(self):
+            if value["class"] == "File":
+                # This is going to re-file -> HDA this each iteration I think, not a good
+                # implementation.
+                return progress.raw_to_galaxy(value)
```
This is nothing right? How did we let this in 😆. It seems it hasn't caused any harm yet!
